cmd/cstor-volume-mgmt/controller/volume-controller/handler.go

```diff
@@ -503,3 +532,77 @@ func IsOnlyStatusChange(oldCStorVolume, newCStorVolume *apis.CStorVolume) bool {
 	klog.Infof("No status changed for cstor volume : %s", newCStorVolume.Name)
 	return false
 }
+
+func (c *CStorVolumeController) upgrade(cv *apis.CStorVolume) (*apis.CStorVolume, error) {
+	var err error
+	if cv.VersionDetails.Current != cv.VersionDetails.Desired {
+		if !isCurrentVersionValid(cv) {
+			return nil, pkg_errors.Errorf("invalid current version %s", cv.VersionDetails.Current)
+		}
+		if !isDesiredVersionValid(cv) {
+			return nil, pkg_errors.Errorf("invalid desired version %s", cv.VersionDetails.Desired)
+		}
+		path := upgradePath(cv)
+		u := &upgradeParams{
+			cv:     cv,
+			client: c.clientset,
+		}
+		cv, err = upgradeMap[path](u)
+		if err != nil {
+			return nil, err
+		}
+		cv.VersionDetails.Current = cv.VersionDetails.Desired
+		cv, err = u.client.OpenebsV1alpha1().
+			CStorVolumes(cv.Namespace).Update(cv)
+		if err != nil {
+			return nil, err
+		}
+		return cv, nil
+	}
+	return cv, nil
+}
+
+// populateVersion assigns VersionDetails for old cv object
+func (c *CStorVolumeController) populateVersion(cv *apis.CStorVolume) (
+	*apis.CStorVolume, error,
+) {
+	v := cv.Labels[string(apis.OpenEBSVersionKey)]
+	// 1.3.0 onwards new CV will have the field populated during creation
+	if v < v130 && cv.VersionDetails.Current == "" {
+		cv.VersionDetails.Current = v
+		cv.VersionDetails.Desired = v
+		obj, err := c.clientset.OpenebsV1alpha1().CStorVolumes(cv.Namespace).
+			Update(cv)
+
+		if err != nil {
+			return nil, err
+		}
+		klog.Infof("Version %s added on cstorvolume %s", v, cv.Name)
+		return obj, nil
+	}
+	return cv, nil
+}
+
+func isCurrentVersionValid(cv *apis.CStorVolume) bool {
```

Comment:
Instead of repeated code can we accept version and decide whether it is valid or not so that we can reuse functions in csp, cvr, cv
all functions like: isCurrentVersionValid, isDesiredVersionValid and upgradePath

Reply:
Will take this in next PR with spc and csp changes also